external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala

```diff
@@ -0,0 +1,559 @@
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def endpointUrl(url: String): Builder[T] = {
+      endpointUrl = url
+      this
+    }
+
+    /**
+     * Sets the AWS region to construct clients for. Defaults to "us-east-1" if no custom value
+     * is specified.
+     *
+     * @param regionName Name of AWS region to use (e.g. "us-west-2")
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def regionName(regionName: String): Builder[T] = {
+      this.regionName = regionName
+      this
+    }
+
+    /**
+     * Sets the initial position data is read from in the Kinesis stream. Defaults to
+     * InitialPositionInStream.LATEST if no custom value is specified.
+     *
+     * @param initialPosition InitialPositionInStream value specifying where Spark Streaming
+     *                        will start reading records in the Kinesis stream from
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def initialPositionInStream(initialPosition: InitialPositionInStream): Builder[T] = {
+      initialPositionInStream = initialPosition
+      this
+    }
+
+    /**
+     * Sets the KCL application name to use when checkpointing state to DynamoDB. Defaults to the
+     * Spark app name if no custom value is specified.
+     *
+     * @param appName Value to use for the KCL app name (used when creating the DynamoDB checkpoint
+     *                table and when writing metrics to CloudWatch)
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def checkpointAppName(appName: String): Builder[T] = {
+      checkpointAppName = appName
+      this
+    }
+
+    /**
+     * Sets how often the KCL application state is checkpointed to DynamoDB. Defaults to the Spark
+     * Streaming batch interval if no custom value is specified.
+     *
+     * @param interval [[Duration]] specifying how often the KCL state should be checkpointed to
+     *                 DynamoDB.
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def checkpointInterval(interval: Duration): Builder[T] = {
+      checkpointInterval = interval
+      this
+    }
+
+    /**
+     * Sets the storage level of the blocks for the DStream created. Defaults to
+     * [[StorageLevel.MEMORY_AND_DISK_2]] if no custom value is specified.
+     *
+     * @param storageLevel [[StorageLevel]] to use for the DStream data blocks
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def storageLevel(storageLevel: StorageLevel): Builder[T] = {
+      this.storageLevel = storageLevel
+      this
+    }
+
+    /**
+     * Use the default AWS credentials provider chain to authenticate to the Kinesis service. This
+     * is also the default credentials provider used if no custom value is specified.
+     *
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisDefaultCredentials(): Builder[T] = {
```

Comment:
Yeah, I mostly put this here for the sake of completeness. Should be able to remove this if we introduce a credential builder and replace it with a single method that just takes a ```SerializableCredentialsProvider``` instance to use for the service. A user could always just manually pass ```DefaultCredentialsProvider``` if they really want to for some reason.